executor/executor.go

```diff
@@ -427,105 +433,142 @@ func (driver *MesosExecutorDriver) Start() (mesosproto.Status, error) {
-func (driver *MesosExecutorDriver) stop(stopStatus mesosproto.Status) error {
+func (driver *MesosExecutorDriver) _stop(stopStatus mesosproto.Status) error {
 	err := driver.messenger.Stop()
-	defer close(driver.destroyCh)
-	defer close(driver.stopCh)
-
-	driver.setStatus(stopStatus)
-	driver.setStopped(true)
+	defer func() {
+		select {
+		case <-driver.stopCh:
+			// already closed
+		default:
+			close(driver.stopCh)
+		}
+		driver.cond.Broadcast()
+	}()
 
+	driver.status = stopStatus
 	if err != nil {
 		return err
 	}
-
 	return nil
 }
 
 // Abort aborts the driver by sending an 'abortEvent' to the event loop, and
 // receives the result from the response channel.
 func (driver *MesosExecutorDriver) Abort() (mesosproto.Status, error) {
-	if stat := driver.Status(); stat != mesosproto.Status_DRIVER_RUNNING {
-		return stat, fmt.Errorf("Unable to Stop, expecting status %s, but got %s", mesosproto.Status_DRIVER_RUNNING, stat)
+	driver.lock.Lock()
+	defer driver.lock.Unlock()
+	return driver.abort()
+}
+
+func (driver *MesosExecutorDriver) abort() (mesosproto.Status, error) {
+	if driver.status != mesosproto.Status_DRIVER_RUNNING {
+		return driver.status, fmt.Errorf("Unable to Stop, expecting status %s, but got %s", mesosproto.Status_DRIVER_RUNNING, driver.status)
 	}
 
 	log.Infoln("Aborting the executor driver")
-	abortStat := mesosproto.Status_DRIVER_ABORTED
-	return abortStat, driver.stop(abortStat)
+	return mesosproto.Status_DRIVER_ABORTED, driver._stop(mesosproto.Status_DRIVER_ABORTED)
 }
 
 // Join waits for the driver by sending a 'joinEvent' to the event loop, and wait
 // on a channel for the notification of driver termination.
 func (driver *MesosExecutorDriver) Join() (mesosproto.Status, error) {
+	driver.lock.Lock()
+	defer driver.lock.Unlock()
+	return driver.join()
+}
+
+func (driver *MesosExecutorDriver) join() (mesosproto.Status, error) {
 	log.Infoln("Waiting for the executor driver to stop")
-	if stat := driver.Status(); stat != mesosproto.Status_DRIVER_RUNNING {
-		return stat, fmt.Errorf("Unable to Join, expecting status %s, but got %s", mesosproto.Status_DRIVER_RUNNING, stat)
+	if driver.status != mesosproto.Status_DRIVER_RUNNING {
+		return driver.status, fmt.Errorf("Unable to Join, expecting status %s, but got %s", mesosproto.Status_DRIVER_RUNNING, driver.status)
+	}
+	for {
+		select {
+		case <-driver.stopCh: // wait for stop signal
+			return driver.status, nil
+		default:
+			driver.cond.Wait()
+		}
 	}
-	<-driver.stopCh // wait for stop signal
-	return driver.Status(), nil
 }
 
 // Run starts the driver and calls Join() to wait for stop request.
 func (driver *MesosExecutorDriver) Run() (mesosproto.Status, error) {
-	stat, err := driver.Start()
+	driver.lock.Lock()
+	defer driver.lock.Unlock()
+	return driver.run()
+}
+
+func (driver *MesosExecutorDriver) run() (mesosproto.Status, error) {
+	stat, err := driver.start()
 
 	if err != nil {
-		return driver.Stop()
+		return driver.stop()
```

Comment:
this is pretty aggressive. consider that the driver has already been started and then the caller invokes run(). would it not be better to return `(DRIVER_RUNNING, ErrAlreadyRunning)` instead of stopping the driver?